pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -173,10 +173,20 @@ private static Map<String, String> parsePrefixesToRename(TableConfig tableConfig
   @Override
   public GenericRow transform(GenericRow record) {
     try {
+      GenericRow originalRow = record.copy();
```

Comment:
This can be costly as we are always making a copy for every record. Can we push this down and only make copy when needed?

Reply:
Make sense, only copy the fields to unnest here.